### ISSUE TITLE
ns object gone when calling .atExit()

### ISSUE THREAD

**Issue by limlim9999** [danielyxie/bitburner#2071]
## Description
`.atExit()` will throw error when `ns` object passed into it.

## Step to reproduce
1. Create any script end with `.ns`.
2. Copy below code snippet and save.
    ```js
    /** @param {NS} ns **/
    export async function main(ns) {
        ns.atExit((ns) => {
            ns.tprint('script dead, printing...');
        })
    }
    ```
3. Run

## Possible fixes
`ws.env.stopFlag` had been set to `true` before going into `workerScript.atExit()`, should be set only after `workerScript.atExit()` had already been proceed. Removing following line seems to fixed the issue.
- [NetscriptWorker.ts, line 48](../blob/dev/src/NetscriptWorker.ts#L48)
  ```diff
    export function prestigeWorkerScripts(): void {
        for (const ws of workerScripts.values()) {
  -         ws.env.stopFlag = true;
            killWorkerScript(ws);
        }

        WorkerScriptStartStopEventEmitter.emit();
        workerScripts.clear();
    }
  ```

**Comment by darchambault** [comment]
Documentation clearly states `.atExit()` to be NS2 exclusive, meaning it only works in NS1 scripts.

That being said, this seems to seriously handicap the usefulness of this method since using callback functions on script exit is a reasonably advanced functionality, and as such would be at least as useful in NS2 than it is in NS1.

Is there a way to add NS2 support to `.atExit()`?

**Comment by aeons** [comment]
If `NS2 exclusive` means that it does _not_ work in NS2, then it really needs a reformulation. I would assume it means it exclusively works in NS2.

**Comment by dwRchyngqxs** [comment]
The function provided to `atExit()` should not take an argument as [it is not called with an argument](https://github.com/danielyxie/bitburner/blob/dev/src/Netscript/killWorkerScript.ts#L73). In your description of the bug, the argument `ns` of the function shadows the `ns` of `main`. `undefined` has no `tprint` function so it outputs an error as it should. This is the intended behaviour.
Now that this is out of the way, I tried to reproduce the bug with the same code but without the `ns` argument to the callback, it did work.
Se either this issue was never something to fix, or it has been solved in the meantime and should be closed.

**Comment by jrchamp** [comment]
Removing the argument results in this code:
```js
/** @param {NS} ns **/
export async function main(ns) {
    ns.atExit(function () {
        ns.tprint('script dead, printing...');
    });
}
```
This produces the following error on the live web version: a3c599e6 (note: I ran the script on n00dles)

> `Error trying to call atExit for script test-atExit.js on n00dles [object Object]`

If you have a working example of ns.atExit(), please share.

**Comment by dwRchyngqxs** [comment]
At the time I tried it worked, I can't reproduce it on current live version.

**Comment by limlim9999** [comment]
Hi @dwRchyngqxs ,

Seems like you are getting the whole thing wrong.

> The function provided to `atExit()` should not take an argument as [it is not called with an argument](https://github.com/danielyxie/bitburner/blob/dev/src/Netscript/killWorkerScript.ts#L73). In your description to reproduce the bug, the argument `ns` of the function shadows the `ns` of `main`. `undefined` has no `tprint` function so it outputs an error as it should. This is the intended behaviour.

You had misunderstood how `atExit()` work, it was an callback function to call when script exiting, see [Reference 1](https://github.com/danielyxie/bitburner/blob/9ddb1c43796d2c020655bae1052e8ed67075c9e9/dist/bitburner.d.ts#L4370),
[Reference 2](https://github.com/danielyxie/bitburner/blob/9ddb1c43796d2c020655bae1052e8ed67075c9e9/src/NetscriptFunctions.ts#L2278).

When you are using `atExit()`, you will need to pass a function to it, which will then available at `workerScript.atExit()`

**Comment by dwRchyngqxs** [comment]
Hi, I did not say `atExit` doesn't take an argument, but the callback doesn't, which is true and is supported by the link you sent. `ns` in `ns.tprint` in the callback is the argument of the callback and not the `main`'s `ns` object, that's called variable shadowing.
Just to clarify one thing, I agree there is a bug when `ns` is not a parameter of the callback in the current version. I just couldn't reproduce the bug when posting my first message so I assumed it was fixed which could be possible given there's been several updates since the issue was opened. You can ignore the part of the message I crossed out.
The rest of the message is relevant if you want to use the example as a test. My point is that you should correct the example.

**Comment by limlim9999** [comment]
> Hi, I did not say `atExit` doesn't take an argument, but the callback doesn't, which is true and is supported by the link you sent. `ns` in `ns.tprint` in the callback is the argument of the callback and not the `main`'s `ns` object, that's called variable shadowing. Just to clarify one thing, I agree there is a bug when `ns` is not a parameter of the callback in the current version. I just couldn't reproduce the bug when posting my first message so I assumed it was fixed which could be possible given there's been several updates since the issue was opened. You can ignore the part I crossed out. The rest of the message is relevant if you want to use the exemple as a test. My point is that you should correct the exemple.

This is exactly why I had opened this issue, "ns object gone when calling .atExit()", can't you see that?
There is no point to repeat over what's other had said.

```js
export async function main(ns) {
	console.log(typeof ns); // object

	ns.atExit((ns) => {
                // if ns is undefined at this point, then there is nothing else we can do to interact with the game
		console.log(typeof ns); // undefined
	});

	ns.atExit(() => {
		// yes, this does work, but you can't interact with the ns object, what's the point? use console.log('script end')?
		console.log(typeof ns); // object
	});
}
```

We should end this conversation at this point, I don't wanted this to go into a pointless argument.
I open this issue is for others to contribute, not for arguing. Hence why I already made the "Possible Fix" section there.

**Comment by dwRchyngqxs** [comment]
In the second case you can't interact with `ns` and that's the point I'm making, this is the bug we're trying to solve (there is already a PR for that #2925). We won't solve the issue by providing `ns` to the callback as done in the first call to `atExit` in your example. This is not the expected behaviour of `atExit` so `ns` being `undefined` is not a bug. Solving the issue could be done by providing `ns` as argument to the callback but neither your fix nor the links you gave point in that direction.

**Comment by Snarling** [comment]
This should be fixed by #3978 which has been merged, I believe this issue can be closed now.
